# Assignment 3: Pandas Part 1


## Learning Objectives
This lesson meets the following learning objectives:
- The ability to use Python data structures provided in Pandas

## Instructions
Read through all of the text in this page. This assignment provides step-by-step training divided into numbered sections. The sections often contain embeded exectable code for demonstration.  Section headers with icons have special meanings:  

- <i class="fas fa-puzzle-piece"></i> The puzzle icon indicates that the section provides a practice exercise that must be completed.  Follow the instructions for the exercise and do what it asks.  Exercises must be turned in for credit.
- <i class="fa fa-cogs"></i> The cogs icon indicates that the section provides a task to perform.  Follow the instructions to complete the task.  Tasks are not turned in for credit but must be completed to continue progress.

Review the list of items in the **Expected Outcomes** section to check that you feel comfortable with the material you just learned. If you do not, then take some time to re-review that material again. If after re-review you are not comfortable, do not feel confident or do not understand the material, please ask questions on Slack to help.

Follow the instructions in the **What to turn in** section to turn in the exercises of the assginment for course credit. 

## Background

This notebook is based on the official Pandas [documentation](https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html).  Pandas is a Python package that we will use to manage labeled data. It has some similarities to R-like DataFrames.  Unless otherwise credited, quoted text comes from the official [10 Minutes to Pandas](https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html#min) documentation which is an excellent resource, and provides the guide which we will drawn from for this tutorial.

*Where does this tool fit in my life?*

> For data scientists, working with data is typically divided into multiple stages: munging and cleaning data, analyzing / modeling it, then organizing the results of the analysis into a form suitable for plotting or tabular display. pandas is the ideal tool for all of these tasks.


*Why use this package?*

> pandas is a Python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive. It aims to be the fundamental high-level building block for doing practical, real world data analysis in Python

### Pandas Cheat Sheet

Cheat sheets to remind you of popular Pandas functionality are available online.  You can find cheat sheets here:

+ [Data Quest Pandas cheat sheet](./media/pandas-cheet-sheet.pdf)
+ [Data Camp cheat sheet](./media/pandas-cheet-sheet.pdf)



## <i class="fa fa-cogs"></i> Notebook Setup

First, we need to import the pandas library (and NumPy library too).  All packages are imported at the top of the notebook. Execute the code in the following cell to get started with this notebook.

In [1]:
# Numpy and pandas usage are often intertwined.
# These abbreviations are ubiquitiously used.
import numpy as np
import pandas as pd

The code above imports both NumPy and Pandas as variables named `np` and `pd`. We can use these variables to access the functionality of both libraries respectively.  Use of `np` and `pd` is commonly used by many. You will find online tutorials and documentation that use this convention. Therefore, we will use those object names for the rest of this class.

You may be wondering why we didn't import pandas like this:  
```python
import pandas
```
We could, but the first is far more commonly seen and the latter would require that we prefix all calls to pandas fuction with `pandas.`. It is easier to just prefix with `pd.`. 

## 1. Data Object Creation

Here we will learn to create and use the two data structures provided by Pandas: **Series** and **DataFrames**.  For additional help, see the [Data Structure Intro section](https://pandas.pydata.org/pandas-docs/stable/getting_started/dsintro.html#dsintro).


In short, Pandas provides two new objects in which to store and manipulate data.

Dimensions | Name |	Description
-----------|-------|--------------------------------------
1 |	Series | 	1D labeled homogeneously-typed array
2 |	DataFrame | 	General 2D labeled, size-mutable tabular structure with potentially heterogeneously-typed column

Typical usage of Pandas with data analysis will involve the creation of one or more `DataFrame` objects, and manipulating them with the provided methods.

### 1.1 The `pd.Series` Data Object

In Pandas, a Series is a one dimensional array, and similar to NumPy, all the values are of the same data type (i.e. integer, string, boolean, etc.) or **NaN**.  NaN means "not a number" but serves as a marker when a value is missing.  

Take a moment to review the [Series online documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html). DataFrames are Python objects, and as such they have member variables and functions. Take a moment to explore the member variables and functions listed on the left hand side of the page in the online documentation. Pick a few and see what they do.  Specifically, take a look at these member variables:  `.shape`, `.dtype`; and these functions: `.astype()`, `.min()`, and `.max()`.


To begin, the following cell creates a new series containing an array of 5 numbers and one missing value by simply passing a list of 6 elements to the `pd.Series` function. Execute the cell:

In [2]:
# Create the series object.
my_series = pd.Series([1, 3, 5, np.nan, 6, 8])

Notice that we have introduced a missing values by using the NumPy variable `np.nan`. 

Execute the next cell to see what the new series object named `my_series` contains

In [ ]:
# Call the object in the notebook.
# This (mostly) calls the __repr__() function of a given object.
my_series

### 1.3. <i class="fas fa-puzzle-piece"></i> Practice

In the cell below notebook, perform the following.

+ Create a series of your own design.

In [26]:
nin_series = pd.Series(np.random.randint(100, size = 100))
nin_series

0     31
1     82
2     99
3      3
4     96
      ..
95    90
96    76
97    31
98    92
99    87
Length: 100, dtype: int32

### 1.4. The `pd.DataFrame` Data Object

In Pandas, a DataFrame is a two-dimensional array.  Just like NumPy, it has axes labeled "rows" and "cols".  However, unlike NumPy, Pandas DataFrames have a few advantages over Numpy 2D matrices:  
1. DataFrames allows for different data types in each column, where as NumPy arrays must always be of the same data type.
2. DataFrames provide greater flexibilty for indexing and slicing.

Take a moment to review the [DataFrames online documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html). DataFrames are Python objects, and as such they have member variables and functions. Take a moment to explore the member variables and functions listed on the left hand side of the page in the online documentation. Pick a few and see what they do.  Specifically, take a look at these member variables:  `.shape`, `.dtypes`; and these functions: `.isna()`, `.duplicated()`, and `.nunique()`.

#### 1.4.1 Create a DataFrame with a dictionary

A dataframe is created by passing a Python 2D dictionary (i.e. **dict**) to the `pd.DataFrame` function.  Remember dictionaries have key/value pairs. The dictionary is interpreted in the following way:

- The keys in the dictionary argument becomes the column names
- The value in the dictionary argument contains the column values
- Each "column" must have the same number of values.

The following cell contains code that creates a new DataFrame object named `df_1`.  This dataframe consists of two columns, one named "alpha" and the other "beta".  Each column has 5 elements. Notice that the columns contain different data types.  Execute the following cell.

In [28]:
# Creating a dataframe with a dictionary.
df_1 = pd.DataFrame(
    {'alpha': [0, 1, 2, 3, 4],
     'beta': ['a', 'b', 'c', 'd', 'e']})

Execute the next cell to see what the new data frame object named `df_1` contains

In [29]:
# Call the object in the notebook.
# This time we get a html output, as the notebook shell
# sees a _repr_html_() function to call.
df_1

alpha beta
0      0    a
1      1    b
2      2    c
3      3    d
4      4    e

Observe that a DataFrame can be visualized as a data table. Importantly, note that, a numerical index has been automatically generated and is shown in the left-most column.

#### 1.4.2. Create a DataFrame with a Numpy array
Alternatively, we can create a DataFrame by passing in a Numpy matrix.  Remember that Numpy arrays can only store one data type, so if the data contains mixed types you would not use this approach. 

The following code creates a Dataframe of 6 rows and 4 columns and fills it with random numbers generated using the Numpy `np.random.randn` function.  Execute the following cell:

In [30]:
df = pd.DataFrame(np.random.randn(6, 4))

Let's take a look at our new data frame by executing the following cell:

In [31]:
df

0         1         2         3
0  1.614016 -1.771160 -0.323967 -0.662389
1 -2.068940  1.078861  0.268372  1.649659
2  0.831256  1.070249  0.456667  0.269318
3  3.094455  0.336472  0.512400  0.534414
4  0.275688  0.348060  0.555119  0.408289
5 -0.701507  0.357057 -2.596331  0.370327

Observe that because we did not provide a dictionary, we do not have header names for each column. Rather a numeric column index was automatically generated.

### 1.5. <i class="fas fa-puzzle-piece"></i> Practice

In the cell below notebook, perform the following.

+ Create a pd.DataFrame object from a Python dictionary. Design the data as you like.

In [36]:
df = pd.DataFrame(np.random.randint(100,size = (4,5)))
df

0   1   2   3   4
0  50  73  34  75  74
1  68  63  62  79   3
2  54  40  29   3  62
3   4  76  38  67   8

### 1.6. Create a DataFrame with row and column labels

Rather than have a numeric index for each row and column of the DataFrame, we may want to give each row a human-readable label. This will help make the data more accessible (as will be seen later).  To demonstrate, lets reuse the 6x4 datafame we created in section 2.2.2. Using this dataset we want to rename the row indexes with hypothetical dates that each row of information was collected, and we want to give the columns names, such as `A1`, `A2`, `B1` and `C1`.  

First, let's set the row index names.  The code below uses the `pd.date_range` function to create a list of dates. Most likely, if we have dates in a real input dataset we would not be creating a list of dates, but for the purposes of demonstration we will generate a series of regularly spaces dates over a 6 day period: one for each of the 6 rows. Note that the `start` argument for the `pd.date_range` function expects the date in YYYYMMDD format.

In [38]:
dates = pd.date_range('20220301', periods=40)
dates

DatetimeIndex(['2022-03-01', '2022-03-02', '2022-03-03', '2022-03-04',
               '2022-03-05', '2022-03-06', '2022-03-07', '2022-03-08',
               '2022-03-09', '2022-03-10', '2022-03-11', '2022-03-12',
               '2022-03-13', '2022-03-14', '2022-03-15', '2022-03-16',
               '2022-03-17', '2022-03-18', '2022-03-19', '2022-03-20',
               '2022-03-21', '2022-03-22', '2022-03-23', '2022-03-24',
               '2022-03-25', '2022-03-26', '2022-03-27', '2022-03-28',
               '2022-03-29', '2022-03-30', '2022-03-31', '2022-04-01',
               '2022-04-02', '2022-04-03', '2022-04-04', '2022-04-05',
               '2022-04-06', '2022-04-07', '2022-04-08', '2022-04-09'],
              dtype='datetime64[ns]', freq='D')

Now, we can re-create the DataFrame from section 2.2.2 with dates as row indexes and  `A1`, `A2`, `B1` and `C1` as column names. We do this by using the `index` and `columns` arguments when creating the DataFrame as shown in the following cell.  Execute the cell to see the results: 

In [41]:
df = pd.DataFrame(np.random.randn(40, 4), index=dates, columns=('A1','A2', 'B1', 'C1'))
df

A1        A2        B1        C1
2022-03-01 -0.629632  1.347608  0.890095 -1.214798
2022-03-02 -0.828762  0.335021 -2.251354  0.341426
2022-03-03 -0.579424  0.470047 -0.372449  0.448439
2022-03-04  1.795538 -0.389218 -0.004423 -0.729677
2022-03-05 -0.710080 -1.787639  0.699401  1.391470
2022-03-06 -1.346271 -0.662713  1.054497  0.278592
2022-03-07  1.518203 -2.936608  1.335732  0.736728
2022-03-08  1.159050  1.800620  0.986842  1.773417
2022-03-09 -0.002733 -1.092928  0.890177  0.265790
2022-03-10  0.297305 -0.181364  0.207630  0.324195
2022-03-11  0.481941 -0.488151  1.108945  1.589813
2022-03-12 -0.214407 -0.370188  0.938141 -0.265592
2022-03-13 -0.976757  0.889948 -1.620689 -0.706614
2022-03-14 -1.307706  0.539332  1.392523 -0.047817
2022-03-15 -0.383208  0.336548 -0.008319 -0.500371
2022-03-16  0.019209  0.912619  0.346484 -0.096851
2022-03-17  0.333505  0.085917  0.233457 -0.044268
2022-03-18  1.863834 -0.219798  0.556713  1.678468
2022-03-19  0.428813 -0.756384  1.254042  0.542603
2022-03-20 -1.860198 -0.942594 -0.571817 -0.437304
2022-03-21 -1.053557  0.879532 -0.131011 -0.243883
2022-03-22  0.136699  0.172910 -1.854487  1.963552
2022-03-23  2.617422 -0.056141 -0.784166 -0.452117
2022-03-24  1.495863 -0.482969  0.550101 -0.109062
2022-03-25 -1.927908 -0.438010 -0.221661 -0.072485
2022-03-26  0.210053  0.078296  0.059271 -1.002340
2022-03-27  0.894084 -1.556082 -0.520454 -0.861633
2022-03-28 -0.286686  0.395883  0.037496  1.818838
2022-03-29  1.935441  0.066970  0.303824  0.354923
2022-03-30  0.771517  0.093581 -1.805502 -1.169552
2022-03-31  1.961290 -0.645971  0.028244  0.998016
2022-04-01 -0.195543  0.536032  0.389814 -0.012168
2022-04-02  0.360462 -0.468653 -0.694263  0.132549
2022-04-03  0.385328  0.418345 -0.767254  0.766508
2022-04-04 -0.407837  0.020291 -1.382524  1.335572
2022-04-05  1.047869  0.351689 -0.642321  1.406272
2022-04-06 -1.053037  0.107302 -2.173910 -0.986469
2022-04-07  1.537787  0.103169 -0.173333  1.170519
2022-04-08  0.876699  1.722083  1.100427  0.125376
2022-04-09 -0.998426 -1.780040 -2.220118  0.750292

### 1.7. <i class="fas fa-puzzle-piece"></i> Practice

In the cell below notebook, perform the following.

+ Create a 10x5 dataframe of random numeric integers that follow a [Guassian (normal) Distribution](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.normal.html). 
  + Center the distrubtion at 0.85.
  + We will use these values as assumed grades for a class of students
+ Adjust the row indexes to be the names of hypothetical students.
+ Adjust the columsn to be the names of hypothetical projects, homework, exam names, etc.


In [52]:
norm = pd.DataFrame(np.random.normal(loc=.85,size=(3,4)), index = ('bob','rob','blob'), columns = ('homework','exam1', 'exam2', 'final exam',))
norm

homework     exam1     exam2  final exam
bob   0.563132  1.265494  0.359065    1.098025
rob   0.666957  0.761830  0.849420    0.575250
blob  0.614028 -0.152104  1.677877    1.501658

## 2. Loading Data

Creating a DataFrame manually, as in the previous examples, would be tedious for large datasets.  Fortunately, Pandas provides a variety of functions for importing data from files.  This saves you from writing your own Python code to open a file, read in the contents and format it for inclusiong into a DataFrame. For this tutorial we will look at two of these import functions:  `pd.read_csv` and `pd.read_excel`.

For this portion of the tutorial, we will use a rather [famous set of data concerning iris flowers](https://en.wikipedia.org/wiki/Iris_flower_data_set).  You can find a copy of the comma-separated file accompanying this notebook in the `data` directory. You can view the file [here](./data/iris.csv).

![Iris flower](./media/A03-193px-Iris_versicolor_3.jpg)


### 2.1 Importing Using `pd.read_csv`

The `pd.read_csv` function can be used to read in files that are in comma-separated format or tab delimited format: two very common data formats.  This function is very flexible and has a large number of arguments. See the [`pd.read_csv` online documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) for a thorough listing of arguments. 

The iris data is in CSV format and is compatible with the default arguments for the `pd.read_csv` function, therefore we can easily import this data by providing the path to the file as the only argument to the function. This will automatically create a DataFrame for us! Execute the following line to import the data:

In [54]:
iris_df = pd.read_csv('data/iris.csv')
iris_df

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

### 2.2. <i class="fas fa-puzzle-piece"></i> Practice

In the cell below notebook, perform the following.

+ Import the iris dataset.
+ Take a look at the `pd.read_csv` online documentation. Write example code in a Markup cell for how you would import this file if it were tab-delimited.


In [ ]:
iris_tab_df = pd.read_csv('data/iris.csv', sep='\t')
iris_tab_df

### 2.3. Importing using `pd.read_excel()`
It is common for data to be stored in Excel worksheets (provided the data is not too large).  You can easily export data from Excel to CSV or Tab delimited formats but Pandas provides  the `pd.read_excel` function to save you the time.  We will not practice loading data from Excel, but take a look at the [`read_excel` online documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html) for details.

## 3. Explore your Data
There are a variety of functions for exploring data once you have imported it. Here we will discuss `pd.head`, `pd.tail`, `pd.sample`, and `pd.describe`.

### 3.1. The `pd.head` function

As with any Python variable, you can display its contents in a Jupyter notebook or on the interactive Python terminal by simply typing the variable name and pressing the "enter" key.  However, for very large data files this may not be desired.  The data used in this tutorial is not extremly large but for large data we need better ways to examine the data.

To easily examine the first set of rows in the DataFrame, use the `head` function of the `pd.DataFrame` object. By deafult it displays the first 5 rows. Execute the following cells to view the first five rows of the iris dataset.

In [55]:
iris_df.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

This is an excellent way to check if you created your `pd.DataFrame` object correctly.  You can change the number of rows returned by providing an argument.  Note, in the table displayed above that the row index have been generated as numeric values but the column indexes are the names of the column headers from the input data file.  If the file has headers, the `pd.read_csv` function will automatically use those as column index names.


The following requests the first 10 rows of the data frame:

In [56]:
iris_df.head(10)

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa
5           5.4          3.9           1.7          0.4  setosa
6           4.6          3.4           1.4          0.3  setosa
7           5.0          3.4           1.5          0.2  setosa
8           4.4          2.9           1.4          0.2  setosa
9           4.9          3.1           1.5          0.1  setosa

### 3.2 The `pd.tail` function
Simliar to the `pd.head` function, the `pd.tail` function lets you peak at the last 5 rows of data.  To demonstrate, execute the following cell.

In [57]:
iris_df.tail()

sepal_length  sepal_width  petal_length  petal_width    species
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

### 3.3 The `pd.sample` function
The `pd.sample` function selects a random set of rows to display. By default it only selects 1 row, we can provide the number of rows we would like (e.g. 5) by providing the number as an arugment:

In [58]:
iris_df.sample(5)

sepal_length  sepal_width  petal_length  petal_width     species
91            6.1          3.0           4.6          1.4  versicolor
136           6.3          3.4           5.6          2.4   virginica
79            5.7          2.6           3.5          1.0  versicolor
73            6.1          2.8           4.7          1.2  versicolor
124           6.7          3.3           5.7          2.1   virginica

Repeat execution of the function to see a different set of randomly selected rows:

In [60]:
iris_df.sample(5)

sepal_length  sepal_width  petal_length  petal_width     species
141           6.9          3.1           5.1          2.3   virginica
34            4.9          3.1           1.5          0.1      setosa
92            5.8          2.6           4.0          1.2  versicolor
64            5.6          2.9           3.6          1.3  versicolor
112           6.8          3.0           5.5          2.1   virginica

Note, the `sample` function can be very important beyond just viewing random rows. If you ever need to select a random sample of records from a dataset the `sample` function can do that for you.

### 3.4 The `pd.describe` function
The `pd.describe` function helps demonstrate the power of Pandas over Python dictionaries and Numpy arrays.  It provides a summary of statistics for the data frame in one simple function call. These statistics include the count, mean, standard deviation, min, max and quartiles.

Execute the following cell to demonstrate:

In [59]:
iris_df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

### 3.5. <i class="fas fa-puzzle-piece"></i> Practice

In the cell below notebook, perform the following.

 + Use `head`, `tail` and `sample` with the iris dataset.
 + Do the same with the dataset you created in task 2c.


## 4. Accessing Data

In this section we will learn to retrieve values, **slice** the data frame, and use column and row labels (rather than numeric indexes) to find data. 

### 4.1 Viewing column and row labels.
Before we delve into retrieving data, lets first look at some properties of DataFrames for retrieving the column and row index names.  It is convenient to work with, or check the indexes and columns of a data frame using their names rather than a numeric index.  First, to retrieve the columns names use the `columns` property of the DataFrame.

In [ ]:
iris_df.columns

Similary, we can get a description of the index using the `index` property:

In [ ]:
iris_df.index

Observe that here our index is numeric and runs from 0 to 150, incremented by 1

### 4.2. <i class="fas fa-puzzle-piece"></i> Practice

In the cell below notebook, perform the following.

+ Display the columns and indexes of the iris dataset.
+ Do the same with the dataset you created in Task 2c.


### 4.3. Get all the values

Perhaps you would like to extract the contents of the DataFrame into a 2-dimensional Python list. You will lose the power of Pandas, but you should know how to do this. There are two ways to extract values: the newer method in Pandas versions above v0.24 way and the older way.  

This following code example used to be the standard way of getting the values of a dataframe as a NumPy array:

```python
my_array = df.values
```

This is depreciated in version 0.24 of Pandas to instead require the use of `to_dict` function.

```python
my_array = df.to_dict()
```

You can verify what version of `pandas` you have:

In [ ]:
pd.__version__

### 4.4. <i class="fas fa-puzzle-piece"></i> Practice

In the cell below notebook, perform the following.

+ Check the version of `pandas` you have 
+ Use the appropriate method to convert the iris data to a dictionary.
+ Do the same with the dataset you created in Task 2c.


### 4.5 Basic Selections
Pandas provides a variety of row/column selectors that are similar to the way that lists and arrays are indxed in Python and NumPy. This tutorial will introduce those. 


#### 4.5.1 Get a column
The easiest way to select a single column is to use the column label! When in doubt, use the `.column` property of the DataFrame to find the list of column labels.  Using the column label will result in the column being returned as a Series object.  For example, if we want the `sepal_length` column of the iris data frame we can execute the following:

In [ ]:
slength = iris_df['sepal_length']
slength.head()

Observe, we can use the `.head` function to get the first few rows of a Series as well as a DataFrame.

#### 4.5.2 Get rows (slicing)
Similar to slicing of Python lists, we can do the same with Pandas DataFrames:

```python
iris_df[start:end]
```

As a reminder, indexing of a list in Python uses 0-based indexing and negative numbers index in reverse order:

```
 +---+---+---+---+---+---+
 | P | y | t | h | o | n |
 +---+---+---+---+---+---+
   0   1   2   3   4   5  
  -6  -5  -4  -3  -2  -1
```

To practice, let's get the first row of the iris dataset by slicing:

In [ ]:
iris_df[0:1]

Next lets get the first five rows:

In [ ]:
iris_df[0:5] 

### 4.4. Selecting with `df.loc` and `df.iloc`

Aside from the column name and slicing (as shown in the previous sections) Pandas provides the `df.loc` and `df.iloc` accessor.  These accessors perform better for selecting subsets of the dataframe. The `.loc` accessor is used to select rows and columns using index names and `.iloc` is used for selecting by numeric indexes.  Although, if your DataFrame only has numeric row indexes then `.loc` can still be used.

However, please note:
> While standard Python / NumPy expressions for selecting and setting are intuitive and come in handy for interactive work, for production code, we recommend the optimized pandas data access methods, '.at', '.iat', '.loc' and '.iloc'.

In other words, rather than index Pandas arrays using something like, `iris[1:5,]`, instead use the functions that will be described below.  They are much faster and optimzied for larger datasets.

#### 4.4.1 Using `.loc`

The `.loc` accessor allows for the slicing of the DataFrame by using the labels of either rows and columns.  

The syntax is as follows:

```python
df.loc[row_start:row_end, column_start:column_end]
```

Remember, because the rows in the iris data frame are indexed using integers, we can use `.loc` to get rows using the integer values.  To get the first row:

In [ ]:
iris_df.loc[0]

To get the first two rows:

In [ ]:
iris_df.loc[0:1]  

Observe that when only one row is returned that a Series is provided.  When multiple rows are returned then a new DataFrame object is returned.


Next, recall that we constructed a 6x4 dataframe in section 2.3 that was filled with random values and we set the row names using datetime values. The DataFrame was named `df`.  As a reminder here are the first 2 rows:

In [ ]:
df.head(2)

In this data frame the rows are indexed using strings that represent dates. We can therefore use those labels to slice the dataframe.  The following extracts the first row:

In [ ]:
df.loc['2022-03-01']

Similarly we can get the first two rows:

In [ ]:
df.loc['2022-03-01':'2022-03-02']

#### 4.4.2 Using `.loc` to get columns
Remember for our iris data frame the column indexes have the names as provided by the `.columns` property.

```python
iris_df.columns
Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width',
       'species'],
      dtype='object')
```

We can use these names to retrive values for specific columns. For example, the code below extracts rows 4 through 6 and the columns 'sepal_width' and 'petal_width':

In [ ]:
iris_df.loc[4:6, 'sepal_width': 'petal_width']

The use of the colon in the code above between 'sepal_width' and 'petal_width' implies a range. Therefore every column between those two are included. This is why 'petal_length' appears in the output.  If we wanted to specifically select only the two columns and none other, we could provide the names in a list:

In [ ]:
iris_df.loc[4:6, ['sepal_width', 'petal_width']]

Observe that now we only have 'sepal_width', and 'petal_width' in the resulting data frame.

Moreover, we can limit the rows to only those specified in the same way:

In [ ]:
iris_df.loc[[4,6], ['sepal_width', 'petal_width']]

#### 4.4.3. <i class="fas fa-puzzle-piece"></i> Practice

In the cell below notebook, perform the following.

+ Use any iris dataframe to:
  + Select a row slice with `loc`.
  + Select a row and column slice with `loc`.
  + Take a look at the [Pandas documentation for the `at` selector](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.at.html). Use what you learn there to select a single item with Pandas `at` accessor.

#### 4.4.4. Using `.iloc`

The `.iloc` accessor allows us to slice the data frame using an integer index, regardless of what the index (or column) labels actually are.

The syntax is the same as `.loc` with the exception of that integers are used instead of labels:

```python
df.iloc[row_start:row_end, column_start:column_end]
```
To retrieve the first row of the iris data we use the numeric index:

In [ ]:
iris_df.iloc[0]  

We can use any combination of integer indexes to get a subset of rows and columns.  The following examples retrieves rows 2 to 5, and excludes the 'species' column.

In [ ]:
iris_df.iloc[2:5, :-1] 

#### 4.4.5.<i class="fas fa-puzzle-piece"></i> Practice

In the cell below notebook, perform the following.


+ Use any iris dataframe to:    
    + Select a row slice with `iloc`.
    + Select a row and column slice with `iloc`.
    + Take a look at the [Pandas documentation for the `iat` selector](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iat.html). Use what you learn there to select a single item with Pandas `iat` accessor.


### 4.5 Boolean Indexing

Pandas allows us to use boolean values to extract specific rows of data that meet certain conditions.To clarify this, lets first examine what occurs when we apply a condition to a data frame.  Create a new dataframe with the following

In [ ]:
df_2 = pd.DataFrame(
    {'alpha': [0, 1, 2, 3, 4],
     'beta': [5, 6, 7, 8, 9]})
df_2

Let's apply a condition to the data frame to find all values > 1.0

In [ ]:
df_2 > 1.0

Notice that the resulting data frame contains `True` and `False` values in place of every value that did or did not meet the condition.  We can use this new data frame to subset the original dataframe.

Lets consider the iris dataset for a more realistic example.  Suppose we want to find all rows with a `sepal_length` greater than 5.8. We know from the call to `describe` that the mean is approximately 5.8. First, we will identify rows that meet our criteria:

In [ ]:
condition = iris_df['sepal_length'] > 5.8
condition

Notice that the Series `condition` indicates `False` if the sepal_length value in the row is <= 5.8 and `True` if it is > 5.8. We can use this array to slice the rows of the dataframe in the following way (only rows with a `True` in the `condition` object are returned):

In [ ]:
iris_df[condition].head()

You could rewrite the lines above to combine them into a single line with identical results:

In [ ]:
iris_df[iris_df['sepal_length'] > 5.8].head()

### 4.6. <i class="fas fa-puzzle-piece"></i> Practice

In the cell below notebook, perform the following.


+ Create subsets of the iris dataset using boolean indexes that:
    + Use one boolean operator.
    + Use two boolean operators.


## Expected Outcomes
At this point, you should feel comfortable with the following:
- What is a Pandas series.
- What is a Pandas dataframe.
- Importing files with Pandas.
- Exploring data with dataframe functions.
- Accession values in dataframes.

## What to Turn in?
Be sure to **commit** and **push** your changes to this notebook.  All practice exercises should be completed.  Once completed, send a **Slack message** to the instructor indicating you have completed this assignment. The instructor will verify all work is completed. 